core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithDecoratorTest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.internal.testing.AnticipatedException;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class RetryingClientWithDecoratorTest {
```

Comment:
A general question unrelated to this PR change:

Say I have 3 decorators on the client as below:

```scala
// Decorator 1
clientBuilder.decorator((delegate, ctx, req) => {
val httpResponse = delegate.execute(ctx, req)
    httpResponse
        .mapHeaders(respHeaders => {
             // throw the created exception X.
        }
})

// Decorator 2
clientBuilder.decorator((delegate, ctx, req) => {
    ctx.log() // ClientRequestContext
          .whenComplete()
          .thenAccept(/* callback Y */)

    val httpResponse = delegate.execute(ctx, req)
        httpResponse
            .mapError(exception => {
               // callback Z: process exception
            })
})

// Decorator 3
clientBuilder.decorator(/* could be any decorator */)
```

Either with or without the decorator 3, after `exception X` is thrown in decorator 1, `callback Z` will always be invoked before `callback Y`, right?

Reply:
> Either with or without the decorator 3, after exception X is thrown in decorator 1, callback Z will always be invoked before callback Y, right?

I believe so in this case.
However, I don't know about your usage, I think it's not a good idea to rely on the execution order of callbacks because callbacks usually do not run in sequence unless you explicitly make it so:
```java
// The execution order of fooCallback and barCallback is different depending on the future's state.
future.handle(fooCallback); 
future.handle(barCallback);

// fooCallback is always executed before barCallback.
future.handle(fooCallback).handle(barCallback)
```

Reply:
Thanks @minwoox for the clarification. 

It makes sense that in your example, `fooCallback` and `barCallback` are added to the same future instance and they may have non-deterministic execution order. However, in my example, `callback Z` is added to `httpResponse` and `callback Y` is added to `ctx.log()`. 

Does that mean Armeria does not guarantee a consistent execution order of callbacks added to different components? IOW, is there no such a contract saying that callbacks added to `ctx.log()` will always be executed before callbacks added to `httpResponse` or vice versa?

The reason why I am asking is because I hope to learn what assumptions I can safely make going forward. Thanks!

Reply:
> callback Z is added to httpResponse and callback Y is added to ctx.log().
Does that mean Armeria does not guarantee a consistent execution order of callbacks added to different components?

No, we don't guarantee that. In your example, if the exception is raised in `mapHeaders`, then Z is called before Y is called. However, if the exception is raised inside and another decorator is attached such as [this](https://github.com/line/armeria/pull/4559#discussion_r1038730448), Y would be called before Z.

>  I hope to learn what assumptions I can safely make going forward. Thanks!

I believe that we shouldn't rely on the execution order of callbacks unless we make it explicitly or there's happens-before relationship between the callbacks.

> callback Z is added to httpResponse and callback Y is added to ctx.log().

There's no happens-before relationship between the callbacks of `HttpResponse` and `ctx.log()`.